File: src/dexter/frontend/clang/src/ir/bool/UnaryExpr.cpp
/*
 * UnaryExpr.cpp
 *
 *  Created on: Mar 24, 2018
 *      Author: maazsaf
 */

#include "ir/bool/UnaryExpr.h"

Dexter::UnaryExpr::Op * Dexter::UnaryExpr::Op::NOT = new Op();
Dexter::UnaryExpr::Op * Dexter::UnaryExpr::Op::PLUS = new Op();
Dexter::UnaryExpr::Op * Dexter::UnaryExpr::Op::MINUS = new Op();
Dexter::UnaryExpr::Op * Dexter::UnaryExpr::Op::BNOT = new Op();

Dexter::UnaryExpr::UnaryExpr (Op * op, Dexter::Expr * base) : Dexter::Expr(NULL)
{
  JNIEnv * e = Util::env;
  jclass c = e->FindClass("dexter/ir/bool/UnaryExpr");
  jmethodID m = e->GetMethodID(c, "<init>", "(Ldexter/ir/bool/UnaryExpr$Op;Ldexter/ir/Expr;)V");
  this->mObj = e->NewObject(c, m, op->obj(), base->obj());
}

void Dexter::UnaryExpr::init ()
{
  JNIEnv * e = Util::env;
  jclass ts = e->FindClass("dexter/ir/bool/UnaryExpr$Op");

  const char * opClass = "Ldexter/ir/bool/UnaryExpr$Op;";

  Op::NOT->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "NOT", opClass));
  Op::PLUS->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "PLUS", opClass));
  Op::MINUS->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "MINUS", opClass));
  Op::BNOT->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "BNOT", opClass));
}

File: src/dexter/ir/bool/UnaryExpr.java
package dexter.ir.bool;

import dexter.ir.Expr;
import dexter.ir.Visitor;
import dexter.ir.type.TypesFactory;

/**
 * Created by Maaz Ahmad on 6/25/19.
 */
public class UnaryExpr extends Expr {

  public static final class Op
  {
    private String symbol;

    private Op(String symbol) { this.symbol = symbol; }

    public static final Op NOT = new Op("!");
    public static final Op BNOT = new Op("~");
    public static final Op MINUS = new Op("-");
    public static final Op PLUS = new Op("+");

    @Override public String toString () { return symbol; }
  }

  protected Op op;
  protected Expr base;

  public UnaryExpr(Op op, Expr base)
  {
    this.op = op;
    this.base = base;

    if (op == UnaryExpr.Op.NOT) { this.type = TypesFactory.Bool; }
    else if (op == UnaryExpr.Op.BNOT) { this.type = base.type(); }
    else if (op == UnaryExpr.Op.PLUS) { this.type = base.type(); }
    else if (op == UnaryExpr.Op.MINUS) { this.type = base.type(); }
  }

  public Op op () { return op; }
  public Expr base () { return base; }

  @Override
  public <T> T accept(Visitor<T> v) {
    return v.visit(this);
  }

  @Override
  public boolean equals (Object o)
  {
    if (o instanceof UnaryExpr)
      return ((UnaryExpr)o).op.equals(op) && ((UnaryExpr)o).base.equals(base);
    else
      return false;
  }
}